mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+    # (a) Initialization: search the 1st source location over the entire
+    # dipoles topographies space
+    # ######################################
+
+    logger.info(' 1st phase : ')
+    s_ap = _fixed_phase1a(attr_dict, data_cov, gain)
+
+    # ######################################
+    # (b) Now, add one source at a time
+    # ######################################
+
+    s_ap = _fixed_phase1b(gain, s_ap, data_cov, attr_dict)
+
+    # #####################################
+    # 2nd phase
+    # #####################################
+
+    logger.info(' 2nd phase : ')
+    s_ap_2 = copy(s_ap)
+    s_ap_2 = _fixed_phase2(attr_dict, s_ap_2, gain, data_cov)
+
+    return s_ap_2
+
+
+def _solve_active_gain_eig(ind, gain, data_cov, eig, perpend_spc):
+    """Eigen values and vector of the projection."""
+    gain_idx = list(range(ind * 3, ind * 3 + 3))
+    l_p = gain[:, gain_idx]
+    eig_a = multi_dot(
+        [l_p.T, perpend_spc, data_cov, perpend_spc, l_p])
+    eig_b = multi_dot(
+        [l_p.T, perpend_spc, perpend_spc, l_p])
+    eig_b = eig_b + 1e-3 * eig_b.trace() * np.eye(3)
+    eig_val, eig_vec = eig(eig_a, eig_b)
+
+    return eig_val, eig_vec, l_p
+
+
+def _free_phase1a(attr_dict, gain, data_cov):
+    """Calculate phase 1a of free oriented AP.
+
+    Initialization: search the 1st source location over
+    the entire dipoles topographies space.
+
+    Parameters
+    ----------
+    attr_dict : dict
+        See: _calculate_free_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, 3 * ndipoles)
+        Gain matrix.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+    oris : array, shape (nsources, 3)
+        Orientations array of estimated sources (sorted by s_ap).
+    sub_g_proj : array
+        Sub space projected by estimated dipoles.
+
+    """
+    from scipy.linalg import eig
+
+    s_ap = []
+    oris = np.empty((attr_dict['nsources'], 3))
+    ap_val1 = np.zeros(attr_dict['ndipoles'])
+    perpend_spc = np.eye(gain.shape[0])
+    for dip in range(attr_dict['ndipoles']):
+        sol_tuple = _solve_active_gain_eig(
+            dip, gain, data_cov, eig, perpend_spc)
+        ap_val1[dip] = np.max([x.real for x in sol_tuple[0]])
+
+    # obtain the 1st source location
+    s1_idx = np.argmax(ap_val1)
+    s_ap.append(s1_idx)
+
+    # obtain the 1st source orientation
+    sol_tuple = _solve_active_gain_eig(
+        s1_idx, gain, data_cov, eig, perpend_spc)
+
+    oris[0] = (
+        sol_tuple[1][:, [np.argmax([x.real for x in sol_tuple[0]])]][:, 0])
+    sub_g_proj = sol_tuple[2] @ oris[0][:, np.newaxis]
+
+    return s_ap, oris, sub_g_proj
+
+
+def _free_phase1b(attr_dict, gain, data_cov,
+                  ap_temp_tuple, force_no_rep):
+    """Calculate phase 1b of free oriented AP.
+
+    Adding one source at a time.
+
+    Parameters
+    ----------
+    ap_temp_tuple : tuple
+        See: _free_phase1a.
```

Comment:
params are not documented in the right order

Reply:
also it's not clear what the See means. You mean in the tuple output of _free_phase1a?

Reply:
Yes, though I'm not sure how to reference it.